src/hardware/mixer.cpp

```diff
@@ -2767,77 +2767,83 @@ void MIXER_CloseAudioDevice()
 	mixer.state = MixerState::Uninitialized;
 }
 
-static bool init_sdl_sound(Section_prop* section)
+// Sets `mixer.sample_rate_hz` and `mixer.blocksize` on success
+static bool init_sdl_sound(const int requested_sample_rate_hz,
+                           const int requested_blocksize_in_frames,
+                           const bool allow_negotiate)
 {
-	const auto negotiate = section->Get_bool("negotiate");
+	SDL_AudioSpec desired  = {};
+	SDL_AudioSpec obtained = {};
 
-	// Start the mixer using SDL sound
-	SDL_AudioSpec spec;
-	SDL_AudioSpec obtained;
+	constexpr auto NumStereoChannels = 2;
 
-	spec.freq     = static_cast<int>(mixer.sample_rate_hz);
-	spec.format   = AUDIO_S16SYS;
-	spec.channels = 2;
-	spec.callback = mixer_callback;
-	spec.userdata = nullptr;
-	spec.samples  = check_cast<uint16_t>(mixer.blocksize);
+	desired.channels = NumStereoChannels;
+	desired.format   = AUDIO_S16SYS;
+	desired.freq     = requested_sample_rate_hz;
+	desired.samples  = check_cast<uint16_t>(requested_blocksize_in_frames);
+
+	desired.callback = mixer_callback;
+	desired.userdata = nullptr;
 
 	int sdl_allow_flags = 0;
 
-	if (negotiate) {
-		// We allow chunk size changes (if supported), but don't allow
-		// frequency changes because we've only seen them go beyond
-		// SDL's desired 48 Khz limit
-#if defined(SDL_AUDIO_ALLOW_SAMPLES_CHANGE)
-		sdl_allow_flags |= SDL_AUDIO_ALLOW_SAMPLES_CHANGE;
-#endif
+	if (allow_negotiate) {
+		// Allow negotiating the sample rate and the audio buffer size
+		// (hopefully to match the audio hardware's native rate to
+		// prevent resampling the final output, and an ideal blocksize
+		// that achieves stutter-free playback at a low latency).
+		//
+		sdl_allow_flags = SDL_AUDIO_ALLOW_SAMPLES_CHANGE |
+		                  SDL_AUDIO_ALLOW_FREQUENCY_CHANGE;
 	}
 
 	constexpr auto SdlError = 0;
 	if ((mixer.sdldevice = SDL_OpenAudioDevice(
-	             nullptr, 0, &spec, &obtained, sdl_allow_flags)) == SdlError) {
+	             nullptr, 0, &desired, &obtained, sdl_allow_flags)) == SdlError) {
+
 		LOG_WARNING("MIXER: Can't open audio device: '%s'; sound output disabled",
 		            SDL_GetError());
 
+		set_section_property_value("mixer", "nosound", "off");
 		return false;
 	}
 
-	// Does SDL want something other than stereo output?
-	if (obtained.channels != spec.channels) {
-		E_Exit("SDL gave us %d-channel output but we require %d channels",
-		       obtained.channels,
-		       spec.channels);
-	}
+	// Opening SDL audio device succeeded
+	const auto obtained_sample_rate_hz = obtained.freq;
+	const auto obtained_blocksize      = obtained.samples;
+
+	mixer.sample_rate_hz = obtained_sample_rate_hz;
```

Comment:
You're doing the correct thing here by setting the mixer to the obtained sample rate (the vast majority of the time, this will also be the same as the requested sample rate).